utils/s2n_mem.c

```diff
@@ -149,6 +149,8 @@ int s2n_dup(struct s2n_blob *from, struct s2n_blob *to)
 {
     eq_check(to->size, 0);
     eq_check(to->data, NULL);
+    ne_check(from->size, 0);
```

Comment:
Might be missing something but if `from->size == 0` should we just `return 0` here?

Reply:
Actually, returning 0 when from->size == 0 causes memory leak in s2n_map.

s2n_dup() calls s2n_alloc() and it calls s2n_realloc(). In case of zero-size realloc, s2n_realloc() returns 0 if s2n_free() succeeds, https://github.com/awslabs/s2n/blob/24568121453cf9504e1564e63cf37b2202594f81/utils/s2n_mem.c#L60

In case of zero-size key, in s2n_map, for example, s2n_map_add() tries to s2n_dup() for key followed by value. https://github.com/awslabs/s2n/blob/24568121453cf9504e1564e63cf37b2202594f81/utils/s2n_map.c#L130 

Since s2n_dup() for key->size == 0 returns 0 although adding key to the map actually fails, it is not caught by GUARD(), then adding value to the map succeeds, which causes unexpected behavior and memory leak.

The fix could also be done by checking &map->table[slot].key after s2n_dup(key, slot) in s2n_map_add/put(). However, I thought checking the source in s2n_dup() would be simpler and cleaner and also I thought it would make sense to return failure in s2n_dup() for zero-size source in that the actual operation would not happen.

Reply:
Makes sense. Simplest solution now is to just not support 0 length keys.